### ISSUE TITLE
snd_pcm_writei: Assertion `size == 0 || buffer' failed

### ISSUE THREAD

**Issue by bz31** [mikebrady/shairport-sync#282]
Hi,

I do not know how to fix this on my Raspberry Pi 3 running piCore.
Thanks for your help.

tc@rpi3:~$ shairport-sync -c /home/tc/shairport-sync.conf -v
Looking for the configuration file "/home/tc/shairport-sync.conf".
Output device name is "default".
Version: "2.8.3-openssl-tinysvcmdns-ALSA"
statistics_requester status is 0.
daemon status is 0.
rtsp listening port is 5000.
udp base port is 6001.
udp port range is 100.
Shairport Sync player name is "Shairport Sync on rpi3".
Audio Output name is "(null)".
on-start action is "(null)".
on-stop action is "(null)".
wait-cmd status is 0.
mdns backend "(null)".
stuffing option is "0".
resync time is 2205.
allow a session to be interrupted: 0.
busy timeout time is 120.
drift tolerance is 88 frames.
password is "(null)".
ignore_volume_control is 0.
disable_synchronization is 0.
audio backend desired buffer length is 6615.
audio backend latency offset is 0.
volume range in dB (zero means use the range specified by the mixer): 0.
zeroconf regtype is "_raop._tcp".
configuration file name "/home/tc/shairport-sync.conf" resolves to "/home/tc/shairport-sync.conf".
New RTSP connection on port 5000
Play connection from user agent "AirPlay/280.32.2".
Active-Remote string seen: "69453147".
Connection from IPv4: 192.168.0.18:52397
PCM handle name = 'default'
alsa device parameters:
  access type = RW_INTERLEAVED
  format = 'S16_LE' (Signed 16 bit Little Endian)
  subformat = 'STD' (Standard)
  number of channels = 2
  number of significant bits = 16
  rate = 44100 frames per second (precisely).
  precise (rational) rate = 0.000 frames per second (i.e. 0/1088784512).
  period_time = 5804 us (>).
  period_size = 256 frames (precisely).
  buffer_time = 5944308 us (>).
  buffer_size = 262144 frames (>).
  periods_per_buffer = 1024 (precisely).
shairport-sync: pcm.c:1251: snd_pcm_writei: Assertion `size == 0 || buffer' failed.
Aborted

**Comment by mikebrady** [comment]
Hi there, and thanks for the report. I've never tried piCore. I just checked on my Pi 3 running Raspbian that it works on the default device, and it's working fine. So, I'd suggest it might be a question for the piCore guys...

**Comment by bz31** [comment]
OK. Thanks.

**Comment by bz31** [comment]
Just for information. 
I have found that piCorePlayer use v2.8.0 https://github.com/ralph-irving/tcz-avahi
I rebuild shairport-sync v2.8.0 (without avahi), it works now !

**Comment by mikebrady** [comment]
That is very interesting, thank you. I must look and see what changes are causing the problem.

**Comment by bz31** [comment]
v2.8.1 works for me too ( ./configure --prefix=/usr/local --with-alsa --with-tinysvcmdns --with-ssl=openssl )

**Comment by plabadens** [comment]
Hi,
I've been getting this exact message on ArchLinux on my RaspberryPi. It only happens when I'm using hardware volume control, and only on versions 2.8.2 and 2.8.3, causing shairport-sync to core dump.
It seems to occur specifically when playing starts, after pausing for instance, or directly on connection.
I tried playing from my Mac and my iPhone, and there seems to be no difference.
As I stated, the problem disappears when I switch to software volume control.
Hopefully this can help pinpointing the problem.

```
Looking for the configuration file "/etc/shairport-sync.conf".
Output device name is "default".
Lowest dB value is a mute.
Can't get dB value corresponding to a "volume" of 1.
Hardware mixer has dB volume from -102.380000 to 4.000000.
Has mute ability.
Version: "2.8.3-openssl-Avahi-ALSA-pulse-soxr"
statistics_requester status is 1.
daemon status is 0.
rtsp listening port is 5000.
udp base port is 6001.
udp port range is 100.
Shairport Sync player name is "Shairport".
Audio Output name is "alsa".
on-start action is "(null)".
on-stop action is "(null)".
wait-cmd status is 0.
mdns backend "avahi".
stuffing option is "0".
resync time is 2205.
allow a session to be interrupted: 0.
busy timeout time is 120.
drift tolerance is 352 frames.
password is "(null)".
ignore_volume_control is 0.
disable_synchronization is 0.
audio backend desired buffer length is 6615.
audio backend latency offset is 0.
volume range in dB (zero means use the range specified by the mixer): 35.
zeroconf regtype is "_raop._tcp".
configuration file name "/etc/shairport-sync.conf" resolves to "/etc/shairport-sync.conf".
avahi: avahi_register.
avahi: register_service.
avahi: request to add "_raop._tcp" service without metadata
avahi: service '2029E6B0BD56@Shairport' successfully added.
avahi: address advertised is: "fe80::821f:2ff:fed3:2fc5".
avahi: address advertised is: "192.168.0.12".
New RTSP connection on port 5000
Closing down RTSP conversation thread...
New RTSP connection on port 5000
Play connection from user agent "iTunes/12.3.3 (Macintosh; OS X 10.11.3)".
Active-Remote string seen: "1823022095".
Connection from IPv4: 192.168.0.14:56545
PCM handle name = 'default'
alsa device parameters:
  access type = RW_INTERLEAVED
  format = 'S16_LE' (Signed 16 bit Little Endian)
  subformat = 'STD' (Standard)
  number of channels = 2
  number of significant bits = 16
  rate = 44100 frames per second (precisely).
  precise (rational) rate = 0.000 frames per second (i.e. 0/1088784512).
  period_time = 5804 us (>).
  period_size = 256 frames (precisely).
  buffer_time = 371519 us (>).
  buffer_size = 16384 frames (>).
  periods_per_buffer = 64 (precisely).
shairport-sync: pcm.c:1251: snd_pcm_writei: Assertion `size == 0 || buffer' failed.
[1]    4884 abort (core dumped)  shairport-sync -v -a Shairport
```

**Comment by mikebrady** [comment]
Many thanks for this. Do you have a reliable sequence of operations to make this happen, by any chance?

**Comment by bz31** [comment]
v2.8.3 works with the usb input of my DAC.  It "Segmentation fault" with my usb interface HiFace2 which has a hardware mixer. See the screenshots of the outputs of alsamixer.
![screenshot at 2016-05-08 09 33 16](https://cloud.githubusercontent.com/assets/1976318/15096632/49793478-1502-11e6-9b69-b3cdc27a1737.png)
![screenshot at 2016-05-07 22 39 55](https://cloud.githubusercontent.com/assets/1976318/15096633/51b69176-1502-11e6-99b1-4cd283be6545.png)

**Comment by mikebrady** [comment]
I've been working on this a bit, though I don't yet know what the problem is. So, I've added extra code to check for either of the conditions apparently causing the assertion violation and to skip the call if such an assertion would fail. It should also give me a clue as to where the fault might be coming from. I've just pushed it into the development branch as 2.8.3.5 and I'd be grateful if you would try it out with log verbosity set to 2.

I suspect that https://github.com/mikebrady/shairport-sync/issues/282#issuecomment-217702017 is a separate problem, so I wouldn't expect it to be changed by these updates...

**Comment by plabadens** [comment]
Now this makes for a rather lengthy log, but here it is. I started and stopped playback, then disconnected and reconnected before starting playback again. Interestingly, the buffer failure happens every time playback starts, except just after connecting.

```
Looking for the configuration file "/etc/shairport-sync.conf".
Looking for configuration file at full path "/etc/shairport-sync.conf"
Output device name is "default".
Open Mixer
Lowest dB value is a mute -- try minimum volume +1
Hardware mixer has dB volume from -102.380000 to 4.000000.
Has mute ability.
The processor is running little-endian.
Version: "2.8.3.5-openssl-Avahi-ALSA-pulse-soxr"
statistics_requester status is 1.
daemon status is 0.
rtsp listening port is 5000.
udp base port is 6001.
udp port range is 100.
Shairport Sync player name is "Shairport".
Audio Output name is "alsa".
on-start action is "(null)".
on-stop action is "(null)".
wait-cmd status is 0.
mdns backend "avahi".
userSuppliedLatency is 0.
AirPlayLatency is -1.
iTunesLatency is -1.
forkedDaapdLatency is -1.
stuffing option is "0" (0-basic, 1-soxr).
resync time is 2205.
allow a session to be interrupted: 0.
busy timeout time is 120.
drift tolerance is 352 frames.
password is "(null)".
ignore_volume_control is 0.
playback_mode is 0 (0-stereo, 1-mono).
disable_synchronization is 0.
audio backend desired buffer length is 6615.
audio backend latency offset is 0.
volume range in dB (zero means use the range specified by the mixer): 35.
zeroconf regtype is "_raop._tcp".
configuration file name "/etc/shairport-sync.conf" resolves to "/etc/shairport-sync.conf".
avahi: avahi_register.
avahi: register_service.
avahi: service '2029E6B0BD56@Shairport' group is not yet commited.
avahi: request to add "_raop._tcp" service without metadata
avahi: service '2029E6B0BD56@Shairport' group is registering.
avahi: service '2029E6B0BD56@Shairport' successfully added.
New RTSP connection from 192.168.0.14:53410 to self at 192.168.0.12:5000.
RTSP connection closed.
Closing down RTSP conversation thread...
RTSP conversation thread terminated.
New RTSP connection from 192.168.0.14:53411 to self at 192.168.0.12:5000.
Play connection from user agent "iTunes/12.3.3 (Macintosh; OS X 10.11.3)".
Active-Remote string seen: "1630075518".
User-Agent is iTunes 10 or better, (actual version is 12); selecting the iTunes latency of -1 frames.
rtp_setup: cport=6001 tport=6002.
Set up play connection from 192.168.0.14 to self at 192.168.0.12.
listening for audio, control and timing on ports 6001, 6002, 6003.
Timing receiver -- Server RTP thread starting.
Control receiver -- Server RTP thread starting.
Audio receiver -- Server RTP thread starting.SET_PARAMETER Content-Type:"text/parameters".
received parameters in SET_PARAMETER request.
volume: -16.000631

Setting volume db to -755.066290.

Timing sender thread starting.
SET_PARAMETER Content-Type:"text/parameters".
received parameters in SET_PARAMETER request.
volume: -16.000631

Setting volume db to -755.066290.
syncing to seqno 92.
PCM handle name = 'default'
alsa device parameters:
  access type = RW_INTERLEAVED
  format = 'S16_LE' (Signed 16 bit Little Endian)
  subformat = 'STD' (Standard)
  number of channels = 2
  number of significant bits = 16
  rate = 44100 frames per second (precisely).
  precise (rational) rate = 0.000 frames per second (i.e. 0/1088784512).
  period_time = 5804 us (>).
  period_size = 256 frames (precisely).
  buffer_time = 371519 us (>).
  buffer_size = 16384 frames (>).
  periods_per_buffer = 64 (precisely).
Open Mixer
Setting volume db to -755.066290.
frame size (fs) < 0!
Zero length silence buffer needed, duh.
Packet reception interval stats: mean, standard deviation and max for the last 2,500 packets in microseconds:     7978.6,     5899.5,    91294.0.
 syncing to seqno 3145.
Open Mixer
Setting volume db to -755.066290.
frame size (fs) < 0!
Zero length silence buffer needed, duh.
frame size (fs) < 0!
Zero length silence buffer needed, duh.
RTSP shutdown requested.
Closing down RTSP conversation thread...
Playback Stopped. Total playing time 00:00:40

Shut down audio, control and timing threads
Audio receiver -- Server RTP thread interrupted. terminating.
Control RTP thread interrupted. terminating.
Timing thread interrupted. terminating.
rtp_timing_sender thread interrupted. terminating.
Closed and terminated timer requester thread.
Timing RTP thread terminated.
timing thread joined
audio thread joined
control thread joined
Player thread exit
shutting down RTP thread
RTSP conversation thread terminated.
New RTSP connection from 192.168.0.14:53414 to self at 192.168.0.12:5000.
Play connection from user agent "iTunes/12.3.3 (Macintosh; OS X 10.11.3)".
Active-Remote string seen: "1630075518".
User-Agent is iTunes 10 or better, (actual version is 12); selecting the iTunes latency of -1 frames.
rtp_setup: cport=6001 tport=6002.
Set up play connection from 192.168.0.14 to self at 192.168.0.12.
listening for audio, control and timing on ports 6001, 6002, 6003.
Timing receiver -- Server RTP thread starting.
Control receiver -- Server RTP thread starting.
Audio receiver -- Server RTP thread starting.
Timing sender thread starting.
SET_PARAMETER Content-Type:"text/parameters".
received parameters in SET_PARAMETER request.
volume: -16.000631

Setting volume db to -755.066290.
SET_PARAMETER Content-Type:"text/parameters".
received parameters in SET_PARAMETER request.
volume: -16.000631

Setting volume db to -755.066290.
syncing to seqno 62142.
Open Mixer
Setting volume db to -755.066290.
RTSP shutdown requested.
Closing down RTSP conversation thread...
Playback Stopped. Total playing time 00:00:25

Shut down audio, control and timing threads
Audio receiver -- Server RTP thread interrupted. terminating.
Control RTP thread interrupted. terminating.
Timing thread interrupted. terminating.
rtp_timing_sender thread interrupted. terminating.
Closed and terminated timer requester thread.
Timing RTP thread terminated.
timing thread joined
audio thread joined
control thread joined
Player thread exit
shutting down RTP thread
RTSP conversation thread terminated.
^Cshutdown requested...
avahi: avahi_unregister.
Request to shut down all rtsp conversation threads
asking playing threads to stop
```

**Comment by mikebrady** [comment]
Thanks for that. Did the software actually work as well, or did it stop? (My guess is that it worked.)

**Comment by plabadens** [comment]
It worked fine, with perhaps minor delays whenever there was an issue with the buffer. Apart from that it kept running until I terminated it.

**Comment by mikebrady** [comment]
Great – the log is giving me a few ideas about where to look.

**Comment by bz31** [comment]
v2.8.3.5 worked for me too (for HiFace2). Many thanks !

tc@rpi3:~$ sudo shairport-sync -c /home/tc/shairport-sync.conf -vv
Looking for the configuration file "/home/tc/shairport-sync.conf".
Looking for configuration file at full path "/home/tc/shairport-sync.conf"
Output device name is "default".
The processor is running little-endian.
Version: "2.8.3.5-openssl-tinysvcmdns-ALSA"
statistics_requester status is 0.
daemon status is 0.
rtsp listening port is 5000.
udp base port is 6001.
udp port range is 100.
Shairport Sync player name is "Shairport Sync on rpi3".
Audio Output name is "(null)".
on-start action is "(null)".
on-stop action is "(null)".
wait-cmd status is 0.
mdns backend "(null)".
userSuppliedLatency is 0.
AirPlayLatency is -1.
iTunesLatency is -1.
forkedDaapdLatency is -1.
stuffing option is "0" (0-basic, 1-soxr).
resync time is 2205.
allow a session to be interrupted: 0.
busy timeout time is 120.
drift tolerance is 88 frames.
password is "(null)".
ignore_volume_control is 0.
playback_mode is 0 (0-stereo, 1-mono).
disable_synchronization is 0.
audio backend desired buffer length is 6615.
audio backend latency offset is 0.
volume range in dB (zero means use the range specified by the mixer): 0.
zeroconf regtype is "_raop._tcp".
configuration file name "/home/tc/shairport-sync.conf" resolves to "/home/tc/shairport-sync.conf".
New RTSP connection from 192.168.0.15:60183 to self at 192.168.0.20:5000.
Play connection from user agent "AirPlay/280.32.2".
Active-Remote string seen: "182122392".
User-Agent is AirPlay; selecting the AirPlay latency of -1 frames.
rtp_setup: cport=51470 tport=52867.
Set up play connection from 192.168.0.15 to self at 192.168.0.20.
listening for audio, control and timing on ports 6001, 6002, 6003.
Audio receiver -- Server RTP thread starting.
Control receiver -- Server RTP thread starting.
Timing receiver -- Server RTP thread starting.
Timing sender thread starting.
SET_PARAMETER Content-Type:"text/parameters".
received parameters in SET_PARAMETER request.
volume: 0.000000

syncing to seqno 14112.
PCM handle name = 'default'
alsa device parameters:
  access type = RW_INTERLEAVED
  format = 'S16_LE' (Signed 16 bit Little Endian)
  subformat = 'STD' (Standard)
  number of channels = 2
  number of significant bits = 16
  rate = 44100 frames per second (precisely).
  precise (rational) rate = 0.000 frames per second (i.e. 0/1088784512).
  period_time = 5804 us (>).
  period_size = 256 frames (precisely).
  buffer_time = 5944308 us (>).
  buffer_size = 262144 frames (>).
  periods_per_buffer = 1024 (precisely).
Setting volume db to 0.000000.
Zero length silence buffer needed, duh.
Zero length silence buffer needed, duh.
frame size (fs) < 0!
Zero length silence buffer needed, duh.
Zero length silence buffer needed, duh.
frame size (fs) < 0!
Zero length silence buffer needed, duh.
frame size (fs) < 0!
Zero length silence buffer needed, duh.
frame size (fs) < 0!
Zero length silence buffer needed, duh.
Zero length silence buffer needed, duh.
frame size (fs) < 0!
Zero length silence buffer needed, duh.
Zero length silence buffer needed, duh.
Zero length silence buffer needed, duh.
Zero length silence buffer needed, duh.
frame size (fs) < 0!
Zero length silence buffer needed, duh.
Zero length silence buffer needed, duh.
Zero length silence buffer needed, duh.
frame size (fs) < 0!
Zero length silence buffer needed, duh.
Zero length silence buffer needed, duh.
frame size (fs) < 0!
Zero length silence buffer needed, duh.
Zero length silence buffer needed, duh.
Zero length silence buffer needed, duh.
frame size (fs) < 0!
Zero length silence buffer needed, duh.
Zero length silence buffer needed, duh.
frame size (fs) < 0!
Zero length silence buffer needed, duh.
Zero length silence buffer needed, duh.
Zero length silence buffer needed, duh.
frame size (fs) < 0!
Zero length silence buffer needed, duh.
frame size (fs) < 0!
Zero length silence buffer needed, duh.
Zero length silence buffer needed, duh.
frame size (fs) < 0!
Zero length silence buffer needed, duh.
frame size (fs) < 0!
Zero length silence buffer needed, duh.
frame size (fs) < 0!
Zero length silence buffer needed, duh.
frame size (fs) < 0!
Zero length silence buffer needed, duh.
frame size (fs) < 0!
Zero length silence buffer needed, duh.
frame size (fs) < 0!
Zero length silence buffer needed, duh.
Zero length silence buffer needed, duh.
Zero length silence buffer needed, duh.
Zero length silence buffer needed, duh.
Zero length silence buffer needed, duh.
Zero length silence buffer needed, duh.
Zero length silence buffer needed, duh.
frame size (fs) < 0!
Zero length silence buffer needed, duh.
Zero length silence buffer needed, duh.
Zero length silence buffer needed, duh.
frame size (fs) < 0!
Zero length silence buffer needed, duh.
Zero length silence buffer needed, duh.
Zero length silence buffer needed, duh.
Zero length silence buffer needed, duh.
frame size (fs) < 0!
Zero length silence buffer needed, duh.
frame size (fs) < 0!
Zero length silence buffer needed, duh.
frame size (fs) < 0!
Zero length silence buffer needed, duh.
Zero length silence buffer needed, duh.
Zero length silence buffer needed, duh.
frame size (fs) < 0!
Zero length silence buffer needed, duh.
Zero length silence buffer needed, duh.
Zero length silence buffer needed, duh.
frame size (fs) < 0!
Zero length silence buffer needed, duh.
frame size (fs) < 0!
Zero length silence buffer needed, duh.
frame size (fs) < 0!
Zero length silence buffer needed, duh.
frame size (fs) < 0!
Zero length silence buffer needed, duh.
frame size (fs) < 0!
Zero length silence buffer needed, duh.
Zero length silence buffer needed, duh.
Zero length silence buffer needed, duh.
frame size (fs) < 0!
Zero length silence buffer needed, duh.
Packet reception interval stats: mean, standard deviation and max for the last 2,500 packets in microseconds:     7980.4,    13263.1,   109191.0.
RTSP shutdown requested.
Closing down RTSP conversation thread...
Shut down audio, control and timing threads
Control RTP thread interrupted. terminating.
Timing thread interrupted. terminating.
rtp_timing_sender thread interrupted. terminating.
Audio receiver -- Server RTP thread interrupted. terminating.
Closed and terminated timer requester thread.
Timing RTP thread terminated.
timing thread joined
audio thread joined
control thread joined
Player thread exit
shutting down RTP thread
RTSP conversation thread terminated.
New RTSP connection from 192.168.0.15:60187 to self at 192.168.0.20:5000.
Play connection from user agent "AirPlay/280.32.2".
Active-Remote string seen: "1809898589".
User-Agent is AirPlay; selecting the AirPlay latency of -1 frames.
rtp_setup: cport=63912 tport=57106.
Set up play connection from 192.168.0.15 to self at 192.168.0.20.
listening for audio, control and timing on ports 6001, 6002, 6003.
Audio receiver -- Server RTP thread starting.
Control receiver -- Server RTP thread starting.
Timing receiver -- Server RTP thread starting.
Timing sender thread starting.
SET_PARAMETER Content-Type:"text/parameters".
received parameters in SET_PARAMETER request.
volume: 0.000000

syncing to seqno 37796.
Setting volume db to 0.000000.
Zero length silence buffer needed, duh.
Zero length silence buffer needed, duh.
Zero length silence buffer needed, duh.
frame size (fs) < 0!
Zero length silence buffer needed, duh.
Zero length silence buffer needed, duh.
frame size (fs) < 0!
Zero length silence buffer needed, duh.
Packet reception interval stats: mean, standard deviation and max for the last 2,500 packets in microseconds:     7979.4,    13743.4,   114124.0.
RTSP shutdown requested.
Closing down RTSP conversation thread...
Shut down audio, control and timing threads
Control RTP thread interrupted. terminating.
Timing thread interrupted. terminating.
rtp_timing_sender thread interrupted. terminating.
Audio receiver -- Server RTP thread interrupted. terminating.
Closed and terminated timer requester thread.
Timing RTP thread terminated.
timing thread joined
audio thread joined
control thread joined
Player thread exit
shutting down RTP thread
RTSP conversation thread terminated.

**Comment by mikebrady** [comment]
Great – I'll clean it up a little for a proper release.

**Comment by mikebrady** [comment]
Fixed in 2.8.4.
